### PR TITLE
Update PyNE to use PyMOAB instead of PyTAPS with newest MOAB

### PR DESCRIPTION
**Pull request by gonuke**
This is a large PR developed collectively by @CNERG to remove the dependency on the deprecated PyTAPS project and replace it with the native python support in MOAB, known as pymoab.

This is the culmination of many months of effort, primarily by @pshriwise, the author of pymoab, and completed by @cnerg over the last month.  

All changes have been internally reviewed, discussed and revised within the @cnerg team.

Details can be found in the news file included in this PR.  Most notably,
* some syntax for mesh and tags has been changed, generally to a more streamlined format
* some dependencies have been more narrowly confined
* new dockerfiles have been added to make it easier to update the CI docker images

Full support for Python 3 should now be possible but is not yet included in this PR.

Any partial review is appreciated and the @cnerg team is, of course, happy to respond to any questions.

This will probably demand a rev of the PyNE version string and a release.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -72,23 +72,44 @@ MESSAGE("--    HDF5 Library directories: ${HDF5_LIBRARY_DIRS}")
 MESSAGE("--    HDF5 Libraries: ${HDF5_C_LIBRARIES}")
 
 
-# Find MOAB library
-if(NOT MOAB_ROOT)
-  set(MOAB_ROOT "${DEPS_ROOT_DIR}")
+# Look for MOAB is requested
```
Comment (katyhuff) [inline CMakeLists.txt]:
```suggestion
# Look for MOAB if requested
```

File: docker/make_pyne_docker_image.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    if args.deps:
+        name += '_pyne-deps'
+    return name
+
+
+def build_docker(args):
+    """Forms the docker command line and executes it. IF the push flag is
+    provided also pushes it to DockerHub which requires PyNE admin right on
+    DockerHub(Docker login is assumed(...
```
Comment (katyhuff) [inline docker/make_pyne_docker_image.py]:
```suggestion
    DockerHub(Docker login is assumed...)
```

File: docker/make_pyne_docker_image.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        ["docker",  "build"] + tag_flag + dockerfile + docker_args + ["."], shell=(os.name == 'nt'))
+
+    if args.push:
+        rtn = subprocess.check_call(
+            ["docker", "push",  tag], shell=(os.name == 'nt'))
+
+
+def main():
+    """Parse the different arguments and call the proper methods.
+
```
Comment (katyhuff) [inline docker/make_pyne_docker_image.py]: delete unnecessary blank line
```suggestion
```

File: docker/make_pyne_docker_image.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    """
+    description = "Build a docker image for PyNE"
```
Comment (katyhuff) [inline docker/make_pyne_docker_image.py]: For consistency
```suggestion
    description = 'Build a docker image for PyNE'
```

File: docker/make_pyne_docker_image.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    all_deps = 'Add all dependencies'
+    parser.add_argument('--all', '-a', '-all', help=all_deps,
+                        action='store_true', default=False)
+
+    deps = 'Depdendencies only'
+    parser.add_argument('--deps', help=deps,
+                        action='store_true', default=False)
+
+    name = "Set docker imgae name"
```
Comment (katyhuff) [inline docker/make_pyne_docker_image.py]: consistency
```suggestion
    name = 'Set docker imgae name'
```

File: docker/make_pyne_docker_image.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    parser.add_argument('--name', help=name, default='')
+
+    push = 'Push docker imgae on dockerhub'
```
Comment (katyhuff) [inline docker/make_pyne_docker_image.py]:
```suggestion
    push = 'Push docker image on dockerhub'
```

File: pyne/dagmc.pyx [hunk 7]
```diff
@@ -368,10 +369,10 @@ def volume_boundary(vol_id):
...
 def point_in_volume(vol_id, xyz, uvw=[1,0,0]):
     """Determine whether the given point, xyz, is in the given volume.
-    
-    If provided, uvw is used to determine the ray fire direction for the underlying 
-    query.  Otherwise, a random direction will be chosen. 
-    
+
+    If provided, uvw is used to determine the ray fire direction for the underlying
+    query.  Otherwise, a random direction will be chosen.
+
```
Comment (katyhuff) [inline pyne/dagmc.pyx]: I think PEP8 discourages extra blank lines, and PEP257 discourages extra blank lines in the docstrings.

```suggestion
```

File: pyne/mesh.py [hunk 8]
```diff
@@ -606,54 +674,61 @@ def __init__(self, mesh=None, structured=False,
...
 
             Structured mesh instantiation:
-                - From iMesh instance with exactly 1 entity set (with BOX_DIMS
-                  tag) by specifying <mesh> and structured = True.
+                - From PyMOAB core instance with exactly 1 entity set
+                  (with BOX_DIMS tag) by specifying <mesh> and structured = True.
                 - From mesh file with exactly 1 entity set (with BOX_DIMS tag)
                   by specifying <mesh_file> and structured = True.
-                - From an imesh instance with multiple entity sets by
+                - From an PyMOAB instance with multiple entity sets by
```
Comment (katyhuff) [inline pyne/mesh.py]:
```suggestion
                - From a PyMOAB instance with multiple entity sets by
```

File: pyne/mesh.py [hunk 9]
```diff
@@ -606,54 +674,61 @@ def __init__(self, mesh=None, structured=False,
...
                 - From coordinates by specifying <structured_coords>,
-                  structured=True, and optional preexisting iMesh instance
-                  <mesh>
+                  structured=True, and optional pre-existing PyMOAB core
+                  instance <mesh>
 
-            The "BOX_DIMS" tag on iMesh instances containing structured mesh is
-            a vector of floats it the following form:
+            The "BOX_DIMS" tag on PyMOAB core instances containing structured
+            mesh is a vector of floats it the following form:
```
Comment (katyhuff) [inline pyne/mesh.py]: Something weird is happening, grammatically here.... maybe this is right:


```suggestion
            mesh is a vector of floats in the following form:
```

File: pyne/mesh.py [hunk 10]
```diff
@@ -666,31 +741,54 @@ def __init__(self, mesh=None, structured=False,
                     raise MeshError("Found {0} structured meshes."
                                     " Instantiate individually using"
                                     " from_ent_set()".format(count))
+
             # from coordinates
             elif (mesh is None) and structured_coords and not structured_set:
+                # check for single vertex coordinates here? it seems we only support volumetric mesh -PCS
```
Comment (katyhuff) [inline pyne/mesh.py:755]: This line seems long. (PEP8 check?)

File: pyne/mesh.py [hunk 11]
```diff
@@ -347,93 +365,138 @@ def __init__(self, size=1, dtype='f8', default=0.0, mesh=None, name=None,
...
+            flat = True if self.size == 1 else False
+            ents = list(miter)[key]
+            data = self.mesh.mesh.tag_get_data(self.tag, ents, flat=flat)
+            return data
         elif isinstance(key, np.ndarray) and key.dtype == np.bool:
             if len(key) != size:
                 raise KeyError("boolean mask must match the length "
                                "of the mesh.")
-            return mtag[[ve for b, ve in zip(key, miter) if b]]
+            return self.mesh.mesh.tag_get_data(self.tag, [ve for b, ve in zip(key, miter) if b], flat=True)
```
Comment (katyhuff) [inline pyne/mesh.py:449]: This line seems long. (PEP8 check?)

File: pyne/mesh.py [hunk 12]
```diff
@@ -1252,23 +1392,86 @@ def tag_cell_fracs(self, cell_fracs):
...
+
+    A StatMesh assumes that each value being operated upon also has a
+    statistical error associaed with it, and forces operations on the
+    statistical error as well.  For any tag with name `tag_name` the StatMesh
+    assumes that there is also a tag with name `tag_name_rel_error`.
+
+    For example, when to quantities are added together, c = a + b, the
+    statistical error of c is found by combining the statistical errors of a
+    and b.
+
```
Comment (katyhuff) [inline pyne/mesh.py]: Unnecessary blank line (see pep8 and pep257)
```suggestion
```

PEP257 recommends one blank line after a class docstring. 

> Insert a blank line after all docstrings (one-line or multi-line) that document a class -- generally speaking, the class's methods are separated from each other by a single blank line, and the docstring needs to be offset from the first method by a blank line.

File: pyne/r2s.py [hunk 13]
```diff
@@ -11,8 +11,65 @@
 warn(__name__ + " is not yet QA compliant.", QAWarning)
 
 
-def irradiation_setup(flux_mesh, cell_mats, alara_params, tally_num=4,
-                      geom=None, num_rays=10, grid=False, flux_tag="n_flux",
+def resolve_mesh(mesh_reference, tally_num = None, flux_tag = "n_flux", output_material=False):
+
```
Comment (katyhuff) [inline pyne/r2s.py]: I don't think a blank line is advisable here. For the docstring to be read properly by many autodoc tools, it really needs to be right after the def line.

```suggestion
```

**Review by katyhuff — APPROVED** [review]
Looks good. I have left a few suggested typo fixes, but otherwise this PR looks very clean. Nice work!

**Comment by katyhuff** [comment]
@pep8speaks suggest diff

File: docker/make_pyne_docker_image.py (hunk 5 above)
Comment (gonuke) [reply docker/make_pyne_docker_image.py]:
```suggestion
    name = 'Set docker image name'
```

File: docker/make_pyne_docker_image.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    args.moab = args.moab or args.dagmc or args.pymoab or args.all
+    args.pymoab = args.pymoab or args.all
+    args.dagmc = args.dagmc or args.all
+
+    return args
+
+
+def build_name(args):
+    """Build the container name, if not specified by the user. Container name
+    will be "pyne/ubuntu_18.04{_moab/_dagmc,_pymoab}{_pyne-deps/}"
```
Comment (makeclean) [inline docker/make_pyne_docker_image.py]: typo , -> /

File: docker/ubuntu_18.04-dev.dockerfile [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+            autoconf \
+            libtool \
+            doxygen \
+    && apt-get clean -y; \
+    if [ "${py_version%.?}" -eq 3 ] ; \
+       then \ 
+            update-alternatives --install /usr/bin/python python /usr/bin/python3 10; \
+            update-alternatives --install /usr/bin/pip pip /usr/bin/pip3 10; \
+    fi;\
+    pip install --force-reinstall \
```
Comment (makeclean) [inline docker/ubuntu_18.04-dev.dockerfile:47]: does pip always point to the right place? I often need to use pip3?

**Review by makeclean — APPROVED** [review]
Guys, this truly is a huge an momentous piece of work - my sincere apologies that its taken me so long to review! When I make it back to Madison I owe you guys beer :)

File: docker/ubuntu_18.04-dev.dockerfile (hunk 15 above)
Comment (gonuke) [reply docker/ubuntu_18.04-dev.dockerfile:47]: I think the above `update-alternatives` will guarantee that for a py3 build

File: docker/make_pyne_docker_image.py (hunk 14 above)
Comment (gonuke) [reply docker/make_pyne_docker_image.py]: This isn't a typo, but maybe it's not clear, the name will be one of:
* `<basename>_moab`
* `<basename>_dagmc`
* `<basename>_dagmc_pymoab`

Possibly `_pyne_deps` will also be appended.

In retrospect, this is an odd combination of options... perhaps we should review.

Comment (gonuke) [reply docker/make_pyne_docker_image.py]:
```suggestion
    will be "pyne/ubuntu_18.04{_moab/_dagmc{_pymoab}}{_pyne-deps/}"
```

Comment (gonuke) [reply docker/make_pyne_docker_image.py]: It'd be great if this didn't hold up the merge and we addressed this in a follow-up

**Comment by gonuke** [comment]
Thanks for the reviews @katyhuff & @makeclean - If someone's up to merge this, we should discuss pushing a new version.

**Comment by makeclean** [comment]
Lets go for the merge :)

**Pull request merged**